components/accelerator/nvidia/error/xid/component.go

```diff
@@ -5,150 +5,240 @@ package xid
 
 import (
 	"context"
-	"fmt"
+	"database/sql"
+	"sort"
+	"strconv"
+	"sync"
 	"time"
 
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
 	"github.com/leptonai/gpud/components"
 	nvidia_common "github.com/leptonai/gpud/components/accelerator/nvidia/common"
 	nvidia_component_error_xid_id "github.com/leptonai/gpud/components/accelerator/nvidia/error/xid/id"
-	nvidia_query_nvml "github.com/leptonai/gpud/components/accelerator/nvidia/query/nvml"
 	nvidia_query_xid "github.com/leptonai/gpud/components/accelerator/nvidia/query/xid"
-	"github.com/leptonai/gpud/components/dmesg"
+	"github.com/leptonai/gpud/components/db"
+	os_id "github.com/leptonai/gpud/components/os/id"
 	"github.com/leptonai/gpud/components/query"
 	"github.com/leptonai/gpud/log"
+	pkg_dmesg "github.com/leptonai/gpud/pkg/dmesg"
 )
 
-func New(ctx context.Context, cfg nvidia_common.Config) components.Component {
+const DefaultRetentionPeriod = 3 * 24 * time.Hour
+const DefaultStateUpdatePeriod = 30 * time.Second
+
+func New(ctx context.Context, cfg nvidia_common.Config, dbRW *sql.DB, dbRO *sql.DB) *XIDComponent {
 	cfg.Query.SetDefaultsIfNotSet()
 	setDefaultPoller(cfg)
 
 	cctx, ccancel := context.WithCancel(ctx)
 	getDefaultPoller().Start(cctx, cfg.Query, nvidia_component_error_xid_id.Name)
 
-	return &component{
-		rootCtx: ctx,
-		cancel:  ccancel,
-		poller:  getDefaultPoller(),
+	setHealthyCh := make(chan struct{})
+	localStore, err := db.NewStore(dbRW, dbRO, "components_accelerator_nvidia_error_xid_events", DefaultRetentionPeriod)
+	if err != nil {
+		log.Logger.Errorw("failed to create store", "error", err)
+		ccancel()
+		return nil
+	}
+	return &XIDComponent{
+		rootCtx:      ctx,
+		cancel:       ccancel,
+		poller:       getDefaultPoller(),
+		setHealthyCh: setHealthyCh,
+		store:        localStore,
 	}
 }
 
-var _ components.Component = (*component)(nil)
-
-type component struct {
-	rootCtx context.Context
-	cancel  context.CancelFunc
-	poller  query.Poller
+func (c *XIDComponent) SetHealthy() error {
+	log.Logger.Debugw("set healthy event received")
+	c.setHealthyCh <- struct{}{}
+	return nil
 }
 
-func (c *component) Name() string { return nvidia_component_error_xid_id.Name }
-
-// Just checks if the xid poller is working.
-func (c *component) States(_ context.Context) ([]components.State, error) {
-	return []components.State{
-		{
-			Name:    StateNameErrorXid,
-			Healthy: true,
-			Reason:  "xid event polling is working",
-		},
-	}, nil
+var _ components.Component = (*XIDComponent)(nil)
+
+type XIDComponent struct {
+	rootCtx      context.Context
+	cancel       context.CancelFunc
+	poller       query.Poller
+	currState    components.State
+	setHealthyCh chan struct{}
+	store        db.Store
+	sync.RWMutex
 }
```

Comment:
It's better to move these before the `New` function. Additionally, giving the `RWMutex` field a name would improve readability.